### PR TITLE
Provide custom execution context for futures

### LINKED ISSUE #617: Provide custom execution context to run futures
We need to implement our own event-loop to run futures. The general idea can stay the same as in Scala.js: future creation pushes the future onto the event loop's queue, as soon execution of the applications' main is complete, event loop starts to pop the futures from the queue and execute them one at a time. As soon as queue is empty, the application terminates.

Default execution context fails to link at the moment with following symbols (as seen for utest): 

```
@java.util.concurrent.BlockingQueue
@java.util.concurrent.ExecutionException
@java.util.concurrent.ExecutionException::init_class.java.lang.String_class.java.lang.Throwable
@java.util.concurrent.Executor
@java.util.concurrent.Executor::execute_trait.java.lang.Runnable_unit
@java.util.concurrent.ExecutorService
@java.util.concurrent.LinkedBlockingQueue
@java.util.concurrent.LinkedBlockingQueue::init
@java.util.concurrent.ThreadFactory
@java.util.concurrent.ThreadPoolExecutor
@java.util.concurrent.ThreadPoolExecutor::allowCoreThreadTimeOut_bool_unit
@java.util.concurrent.ThreadPoolExecutor::init_i32_i32_i64_class.java.util.concurrent.TimeUnit_trait.java.util.concurrent.BlockingQueue_trait.java.util.concurrent.ThreadFactory
@java.util.concurrent.TimeoutException
@java.util.concurrent.TimeoutException::init_class.java.lang.String
@java.util.concurrent.locks.AbstractQueuedSynchronizer
@java.util.concurrent.locks.AbstractQueuedSynchronizer::acquireSharedInterruptibly_i32_unit
@java.util.concurrent.locks.AbstractQueuedSynchronizer::init
@java.util.concurrent.locks.AbstractQueuedSynchronizer::releaseShared_i32_bool
@java.util.concurrent.locks.AbstractQueuedSynchronizer::toString_class.java.lang.String
@java.util.concurrent.locks.AbstractQueuedSynchronizer::tryAcquireSharedNanos_i32_i64_bool
@scala.concurrent.forkjoin.ForkJoinPool
@scala.concurrent.forkjoin.ForkJoinPool$ForkJoinWorkerThreadFactory
@scala.concurrent.forkjoin.ForkJoinPool::execute_class.scala.concurrent.forkjoin.ForkJoinTask_unit
@scala.concurrent.forkjoin.ForkJoinPool::init_i32_trait.scala.concurrent.forkjoin.ForkJoinPool$ForkJoinWorkerThreadFactory_trait.java.lang.Thread$UncaughtExceptionHandler_bool
@scala.concurrent.forkjoin.ForkJoinTask
@scala.concurrent.forkjoin.ForkJoinTask::fork_class.scala.concurrent.forkjoin.ForkJoinTask
@scala.concurrent.forkjoin.ForkJoinTask::init
@scala.concurrent.forkjoin.ForkJoinWorkerThread
@scala.concurrent.forkjoin.ForkJoinWorkerThread::getPool_class.scala.concurrent.forkjoin.ForkJoinPool
@scala.concurrent.impl.AbstractPromise
@scala.concurrent.impl.AbstractPromise::getState_class.java.lang.Object
@scala.concurrent.impl.AbstractPromise::init
@scala.concurrent.impl.AbstractPromise::updateState_class.java.lang.Object_class.java.lang.Object_bool
```

Blocks: #435

Related: https://github.com/scala-js/scala-js/blob/master/scalalib/overrides/scala/concurrent/ExecutionContext.scala and https://github.com/scala-js/scala-js/tree/master/library/src/main/scala/scala/scalajs/concurrent

### PR DESCRIPTION
**Pull request by densh**
Fixes #617 

Supersedes #615 

Review by @sjrd

### CODE REVIEW

File: sandbox/Test.scala [hunk 1]
```diff
@@ -1,4 +1,7 @@
 object Test {
-  def main(args: Array[String]): Unit =
-    println("Hello, world!")
+  def main(args: Array[String]): Unit = {
+    import scala.concurrent.Future
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future(1 + 2).map(_ + 3).map(_ + 4).foreach(println)
+  }
```
Comment (densh) [inline sandbox/Test.scala]: This needs to be upgraded to scripted test before we merge this.

File: scalalib/overrides-2.11/scala/concurrent/ExecutionContext.scala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    /**
+     * The implicit global `ExecutionContext`. Import `global` when you want to provide the global
+     * `ExecutionContext` implicitly.
+     *
+     * The default `ExecutionContext` implementation is backed by a work-stealing thread pool. By default,
+     * the thread pool uses a target number of worker threads equal to the number of
+     * [[https://docs.oracle.com/javase/8/docs/api/java/lang/Runtime.html#availableProcessors-- available processors]].
+     */
+    implicit lazy val global: ExecutionContextExecutor =
+      scala.scalanative.runtime.ExecutionContext.global
```
Comment (densh) [inline scalalib/overrides-2.11/scala/concurrent/ExecutionContext.scala]: That's the only method overriden here.

**Comment by densh** [comment]
Just added a scripted tests that futures are indeed working now. Any more test ideas, @sjrd ?

File: build.sbt [hunk 3]
```diff
@@ -356,6 +356,8 @@ lazy val scalalib =
         IO.copyDirectory(file(s"scalalib/overrides-$epoch.$major/scala"),
                          file("scalalib/src/main/scala/scala"),
                          overwrite = true)
+
+        (file("scalalib/src/main/scala") ** "*.java").get.foreach(IO.delete)
```
Comment (sjrd) [inline build.sbt]: Not sure what this does.

File: scripted-tests/run/execution-context/build.sbt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+ScalaNativePlugin.projectSettings
```
Comment (sjrd) [inline scripted-tests/run/execution-context/build.sbt]: This should ideally be
```scala
enablePlugins(ScalaNativePlugin)
```

File: scripted-tests/run/execution-context/src/main/scala/Hello.scala [hunk 5]
```diff
@@ -0,0 +1,7 @@
+object Test {
+  def main(args: Array[String]): Unit = {
+    import scala.concurrent.Future
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future(1 + 2).map(_ + 3).map(_ + 4).foreach(println)
+  }
```
Comment (sjrd) [inline scripted-tests/run/execution-context/src/main/scala/Hello.scala]: You should also `println("main")` at the end of `main()` (after the `Future(1 + 2)...` line) and assert that the output is `"main\n10"` (and not `"10\nmain"`).

File: build.sbt (hunk 3 above)
Comment (densh) [inline build.sbt]: Removes all the java code from our build of scala library. This fixes conflict situations when we have a Scala-defined override for something that used to be in Java before.

File: scripted-tests/run/execution-context/build.sbt (hunk 4 above)
Comment (densh) [inline scripted-tests/run/execution-context/build.sbt]: Good point.

File: build.sbt (hunk 3 above)
Comment (densh) [inline build.sbt]: I should add comment to explain this.

File: scripted-tests/run/execution-context/src/main/scala/Hello.scala (hunk 5 above)
Comment (densh) [inline scripted-tests/run/execution-context/src/main/scala/Hello.scala]: Good idea.

File: build.sbt (hunk 3 above)
Comment (densh) [inline build.sbt]: Done.

File: scripted-tests/run/execution-context/build.sbt (hunk 4 above)
Comment (densh) [inline scripted-tests/run/execution-context/build.sbt]: Done.

File: scripted-tests/run/execution-context/src/main/scala/Hello.scala (hunk 5 above)
Comment (densh) [inline scripted-tests/run/execution-context/src/main/scala/Hello.scala]: Done.

**Pull request merged**